### ISSUE TITLE
Unable to reproduce the results by running codes from scratch instead of using given prototypes

### ISSUE THREAD

**Issue by Chandan-IITI** [mlbio-epfl/crow#1]
Hi @mbrbic and @wenshuo128 ,

I followed all mentioned four steps. I also used CLIP pretrained model from the mentioned CLIP repository from the following link: https://openaipublic.azureedge.net/clip/models/3035c92b350959924f9f00213499208652fc7ea050643e8b385c2dac08641f02/ViT-L-14-336px.pt

**By using given prototypes**, I am able to reproduce the results. However, I am **unable to reproduce** the results by running (training) **codes from scratch instead of using given prototypes**.

Can you please guide me if I am missing anything and what do I need to do for reproducing the results?

**Comment by wenshuo128** [comment]
Thank you for the question! I sent you an email to ask about some details.

For the other users, I will update the answer and maybe the code after a thorough discussion of this question.

**Comment by Chandan-IITI** [comment]
Thank you for your resaponse. I ran the code over **OfficeHome dataset** and results are drastically different from the results obtained from the given prototype for **unseen classes**. Please see the attachment 

![image](https://github.com/user-attachments/assets/9439df46-2fcd-45ab-a566-f1ca456b91a5)

**Comment by Chandan-IITI** [comment]
> the question! I sent you an emai

Sorry, I didn't recieve any email regarding this. Can you please resend it? 
My email is chandangautam31@gmail.com

**Comment by wenshuo128** [comment]
It's okay. I sent an email asking for some details, and you have already included them here :)

The problem is the model is biased toward the seen classes. We can see that the seen accuracy is much better compared to the number I reported in the paper, but the unseen accuracy is much worse. During fine-tuning, how large was the batch size you set? The L_reg is the key that keeps the balance between seen and unseen classes, and a small batch size or a very large learning rate can break the balance. Please let me know if you used the default batch size I provided in the paper. Then, I will check why the model is biased toward the seen classes. Thank you!

**Comment by Chandan-IITI** [comment]
Thanks for the response. I ran the code without making any hyperparameter change in the code. As given in the instruction of readme file for OfficeHome dataset, we follow the exactly same. I didn't set any specific batch size or even modify a single word in the code and rand it for OfficeHome dataset. 

**In the paper,** it is mentioned **"We set the batch size to 32..."**. I can see four files in the code has batch size as an argument:

**get_features.py :**  dataloader_train = DataLoader(dataset_train, batch_size=1, shuffle=False, num_workers=32, pin_memory=True)
**get_prototype_source.py:**   parser.add_argument("--batch_size", type=int, default=256)
**get_prototype_target.py:** parser.add_argument("--batch_size", type=int, default=1024)
**matching.py:** parser.add_argument("--batch_size", type=int, default=32)

Please let me know if I need to change anything in the code.

**Comment by wenshuo128** [comment]
Thank you for the details! I will reply to you after checking the code.

**Comment by wenshuo128** [comment]
Thank you for the details! I updated 'get_prototype_target.py'. In general, I just remembered that I need to upload the part of supervised learning from the K-means pseudo label. Before, the seen prototypes were obtained from supervised learning, while the unseen prototypes were obtained from K-means; they were not at the same level. Now, it should output normal results (balanced results between seen and unseen) even if the results can be unstable. 

Louvain clustering algorithm can make the results more stable, and I will upload it in the coming days :)

Thank you again for using our code and reporting this problem!

**Comment by Chandan-IITI** [comment]
Thanks a lot @wenshuo128 for updating the code. I tried to rerun the updated code. It significantly improved the results from the earlier but still **significantly lacking from the results mentioned in the paper** (and also from what I obtained from the given prototype).  Please refer this screnshot:

![image](https://github.com/user-attachments/assets/96c982aa-abf2-491a-90fb-77a4f46eda18)

I also rerun the code multiple times but everytime getting similar results as above. **Do I need to change any hyperparameter?** For now, I executed the code without any change for OfficeHome Dataset.

**Comment by wenshuo128** [comment]
Thank you for the results! I will update the 'get_prototype_target.py' by adding the Louvain clustering algorithm as soon as possible. It will lead to more stable and better results on smaller datasets like Office and OfficeHome :) For Office and OfficeHome datasets, k-means can be unstable, and you can play with it a bit by changing the random seed of the k-means algorithm.

**Comment by Chandan-IITI** [comment]
Hi @wenshuo128,

Thank you for the update. I noticed one thing in the code in the file 'get_prototype_target.py'. In the earlier version of this code, KMEANS was **with random state**=8 as follows:

    kmeans = KMeans(n_clusters=total_class_num, random_state=8, n_init="auto").fit(feature_target.cpu())

However, in the current version of the code, it is **without random state** like follows:

    kmeans = KMeans(n_clusters=total_class_num, n_init="auto").fit(feature_target.cpu())

Results with random state=8  are significantly better than without random state. I just wanted to let you know. I hope this issue will be resolved when you update the code with the Louvain clustering algorithm.

**Comment by wenshuo128** [comment]
I updated the Louvain algorithm. I ran the code several times and found that the model can sometimes be slightly biased toward seen classes. However, the H-score should be very similar to the number in the paper (for me, the results are usually a bit better).

Thank you again for using our code and finding the problem!

**Comment by Chandan-IITI** [comment]
@wenshuo128, Thanks a lot for the update. It is really helpful. 

I also rerun the code for OfficeHome and Office datasets. **For the Office dataset**, results are reproducible. However, for the OfficeHome dataset, **the code was stuck in the while loop for the target 'RealWorld.'** 

The code was perfectly executed for other targets, such as art, clipart, and product, but not for the target 'RealWorld'. When we change the line **"sc.pp.neighbors(adata_ori)"** to **"sc.pp.neighbors(adata_ori, use_rep='X')"**, then the loop condition was satisfied, but the results are 4-5% below the mentioned results in the paper. 

Can you please help me with this?

**Comment by wenshuo128** [comment]
Thank you for the update! 

I never met this problem before. Could you please fix a random.seed in 'get_prototype_target.py' and then send me the seed? Thank you!

**Comment by Chandan-IITI** [comment]
@wenshuo128  Thank you for the response and help!

I added the following lines to fix all kinds of randomness (Set random seed for Python, NumPy, and PyTorch) in the file: 'get_prototype_target.py'

    import random
    import numpy as np

    random.seed(42)
    np.random.seed(42)
    torch.manual_seed(42)
    torch.cuda.manual_seed_all(42)  # For CUDA if using GPU

My code is stuck in a while loop for any seed. Please help me resolve this issue.

**Comment by wenshuo128** [comment]
I tried the seed and sc.pp.neighbors(adata_ori, use_rep='X'), here are my findings:

1. With 'use_rep='X'', the function will not automatically apply PCA to 'X', and this really **improves the overall performance** (averagely ~2-3% in H-score for OfficeHome compared to the results in the paper). Thank you so much for pointing this out! I have updated the code and changed it to sc.pp.neighbors(adata_ori, use_rep='X').

2. Even without 'use_rep='X'', I did not encounter the problem using the seed you sent me. I do not know what really happened, but I will provide you with a possible solution.

**Possible solution:**
Unlike K-means, we cannot set how many clusters we need in Louvain. Instead, we need to adapt 'n_neighbors' in sc.pp.neighbors and 'resolution' in sc.tl.louvain to get the number of clusters we need. In general, a higher 'resolution' leads to more clusters. Thus, in the code, I write a self-adaption method to automatically adapt the 'resolution', and for me it always works well. For debugging, I suggest you to do the following:

1. Print 'num_cluster' and 'res' in 'get_prototype_target.py'. Maybe you will see the 'num_cluster' jumps between 64 and 66. 
2. Then, you can try the average of the two corresponding 'res'.
3. If the average 'res' still does not work, you can manually adapt 'res' to get the number of clusters you need (65 for OfficeHome). Since '65 classes' is an input to the algorithm, there is no data leaky problem to use it to adapt this hyperparameter.

Let me know if this can help :)

**Comment by Chandan-IITI** [comment]
@wenshuo128 Thank you for the suggestion. I am not sure what is the issue but I am not even able to get similar performance as  mentioned in the paper while using ''use_rep='X''. I tried to run multiple times. I am getting following results

For Art to Clipart on OfficeHome Dataset:

Seen accuracy - 0.815; Unseen accuracy - 0.515; HScore - 0.6311 **(run by me using  ''use_rep='X'')**
Seen accuracy - 0.729; Unseen accuracy - 0.614; HScore - 0.6665 **(From the given prototype of your github codebase)**

I am not sure why there is such differences and you are getting even better than the paper using ''use_rep='X''. Can it be a package version mismatch?

**Comment by wenshuo128** [comment]
My scanpy and anndata version is: anndata-0.9.2; scanpy-1.9.4

Could you please try the other combinations? Art to Clipart and Clipart to Art are the hardest cases, and it can sometimes have ~2-3% difference in unseen accuracy (better or worse) compared to the average performance I reported in the paper. However, I never met Unseen accuracy - 0.515 before; sometimes I meet Unseen accuracy ~ 0.585.

If it is possible, could you please upload all the performance you get from all the combinations of Office and OfficeHome datasets? With all the results, it is easier for me to find a possible solution.

**Comment by Chandan-IITI** [comment]
@wenshuo128 Thanks a lot for your suggestion. I tried keeping the scanpy and anndata versions as anndata-0.9.2 and scanpy-1.9.4. But still the same results. 

**Is it possible to provide the list of all packages in that environment by using 'conda list' or version of these packages: Pandas, Numpy, SKLearn, Torch, torchvision, and PIL?**

My while loop oscillates in between 64 and 66 but never reaches 65. Therefore, unable to generate results.

**Comment by wenshuo128** [comment]
1. In your comment yesterday, you said that when you changed to sc.pp.neighbors(adata_ori, use_rep='X'), the loop condition was satisfied. You can use this version to generate results :) There should not be a huge difference, so those results are already enough for me to find the possible problem. In general, I want to see if the problem lies in all the combinations or if it is for some specific combinations.
2. The version of the other packages: Pandas-2.1.0, Numpy-1.24.4, SKLearn-1.3.0, Torch-2.0.1, torchvision-0.15.2, PIL-9.4.0.
3. For the problem that your while loop oscillates in between 64 and 66 but never reaches 65, you can try the third step in the possible solution I provided before. Since I can reach 65, there should be a resolution to make the cluster number 65. Even if I do not know why it cannot be found by the self-adaption algorithm from your side, you can still try to manually find it by trying the resolutions between the resolution of 64 and 66 clusters.

**Comment by Chandan-IITI** [comment]
Thanks a lot, @wenshuo128. I tried your suggestion in point number 3, and it works well. The problem only lies with the Art to Clipart combination. I was running code for some other dataset. 

It seems there is **some discrepancy in the listing of source and target_list for VisDA dataset** in all **four** files: 

get_features.py (line numbers 24 and 25)
get_prototype_target.py  (line numbers 50 and 51)
 get_prototype_source.py  (line numbers 35 and 36)
 and matching.py   (line numbers 55 and 56)

**Can you please verify it, as all have different sets of sources and target_list for VisDA dataset, which yields errors in running the code matching.py?**

**Comment by wenshuo128** [comment]
Thank you for the question! For VisDA, we always use 'train' as the source and 'validation' as the target, as in the matching.py. In the other three files, we aim to apply the function to the target_list. For example, we want to get the features of all the samples in both 'train' and 'validation', so the target_list is ['train', 'validation'] in get_feature.py. When we want to get the prototypes of the 'train' set, the target list is ['train'] in the get_prototype_source.py.

I also send you an email (chandangautam31@gmail.com) to further discuss the details :)
